### ISSUE TITLE
Change event fires too much when inputing Chinese characters.

### ISSUE THREAD

**Issue by chenxsan** [facebook/react#3926]
When I was trying this [example](https://jsfiddle.net/reactjs/n47gckhr/light/) from https://facebook.github.io/react/blog/2013/11/05/thinking-in-react.html, any Chinese characters inputted by Chinese pinyin input method would fire too many renders like:

![screen shot 2015-05-21 at 14 04 36](https://cloud.githubusercontent.com/assets/1091472/7742565/2c8625b0-ffc3-11e4-8ac2-d7eb22a3aef3.png)

Actually I would expect those not to fire before I confirm the Chinese character.

Then I tried another kind of input method - wubi input method, I got this:

![screen shot 2015-05-21 at 14 17 15](https://cloud.githubusercontent.com/assets/1091472/7742657/42c27bac-ffc4-11e4-8a3a-7d4550e88a59.png)

It's weird too. So I did a test [in jQuery](http://jsbin.com/yepogahobo/1/edit?html,js,console,output):

![screen shot 2015-05-21 at 14 05 12](https://cloud.githubusercontent.com/assets/1091472/7742591/71da6842-ffc3-11e4-9d7a-a8438721029c.png)

Only after I press the space bar to confirm the character, the `keyup` event would fire.

I know it might be different between the implementation of jQuery `keyup` and react `onChange` , but I would expect the way how jQuery `keyup` handles Chinese characters instead of react's `onChange`.

**Comment by spicyj** [comment]
cc @salier :) – What should we do here?

**Comment by salier** [comment]
I think we should not fire `onChange` until the IME string is committed.

One way to handle this in `ChangeEventPlugin` would be to ignore all `input` events between `compositionstart` and `compositionend`, then use the `input` event immediately following `compositionend`.

I did some quick testing on OSX Chrome and Firefox with Simplified Pinyin and 2-Set Korean, and the event order and data seem correct enough. (I predict that we'll have problems with IE Korean, but we may get lucky.)

I think we may continue to see issues with alternative input methods like the Google Input Tools extension, but there may be workarounds for that.

**Comment by Bertg** [comment]
This also influences how dialectic characters are typed for latin languages. For example on OS X inserting an ```é``` will fail using an international keyboard. Even the long press ```e``` and then use variant are failing here.

**Comment by jasonslyvia** [comment]
Is there any update? Suffering from this issue too.

**Comment by spicyj** [comment]
None currently – this is not a high priority for us right now. I'd be happy to look at a pull request if anyone dives into fixing this.

**Comment by pswai** [comment]
@salier It seems like IE does not fire `input` event after `compositionend`. I have tested on IE11 and Edge on Windows 10. It fires properly in Chrome and Firefox.

**Comment by suhaotian** [comment]
in ie 9, Change event fires too many times when inputing Chinese characters again

**Comment by zhaoyao91** [comment]
Hello, Facebook guys, in fact this issue causes a **SERIOUS** problem: we can't update input asynchronously with Chinese input. 
For example, we can't use meteor reactive datasources or stores like redux, because they all feedback update asynchronously.
Here is a simplest example to show this problem, it use setTimeout to do async update:
[https://jsfiddle.net/liyatang/bq6oss6z/1/](https://jsfiddle.net/liyatang/bq6oss6z/1/)

I really hope you can fix this quickly, so that we won't waste efforts to workaround it here and there and over and over again.

Thanks.

**Comment by eyesofkids** [comment]
I made a simple example to demo how to use `compositionstart` and `compositionend` events to prevent inputing Chinese IME on `onchange` event problem. 
Here is the link: [https://jsfiddle.net/eyesofkids/dcxvas28/](https://jsfiddle.net/eyesofkids/dcxvas28/)

**Comment by zhaoyao91** [comment]
@eyesofkids good work, this could be made as the default input, textarea ... implementation for onChange

**Comment by suhaotian** [comment]
nice work !

**Comment by julen** [comment]
I was hitting the same issue and @eyesofkids' workaround works perfectly (thank you!).

After having the workaround in place, I was diving into React's source code to at least try to add a failing test for this —hoping to later add the expected behavior to the library— although it seems a bit complicated for someone unfamiliar with the internals.

Initially I was expecting that a test similar to what's already available for `ChangeEventPlugin` should work, i.e. simulating a native `compositionStart`/`compositionUpdate` and checking no `onChange` callback was fired; also checking `onChange` would only be fired once `compositionEnd` was simulated. However this doesn't seem to work.

Hence I was thinking that maybe checking for `ChangeEventPlugin.extractEvents()` would be a feasible approach, similar to what's done in the tests for `SelectEventPlugin`. Here for some reason I always get `undefined` when extracting the events though.
For reference, this is the test code I tried within _ChangeEventPlugin-test.js_:

```jsx
  var EventConstants = require('EventConstants');
  var ReactDOMComponentTree = require('ReactDOMComponentTree');
  var topLevelTypes = EventConstants.topLevelTypes;

  function extract(node, topLevelEvent) {
    return ChangeEventPlugin.extractEvents(
      topLevelEvent,
      ReactDOMComponentTree.getInstanceFromNode(node),
      {target: node},
      node
    );
  }

  function cb(e) {
    expect(e.type).toBe('change');
  }
  var input = ReactTestUtils.renderIntoDocument(
    <input onChange={cb} value='foo' />
  );

  ReactTestUtils.SimulateNative.compositionStart(input);

  var change = extract(input, topLevelTypes.topChange);
  expect(change).toBe(null);
```

I'm afraid I don't know exactly how one is supposed to debug these tests—otherwise I'd have a clearer picture of what's going on. Any guidance on how to proceed or any other pointers would be highly appreciated.

**Comment by julen** [comment]
The workaround suddenly broke in Chrome 53+ and it seems it is not valid anymore because [they changed the order `compositionend` is fired](https://chromium.googlesource.com/chromium/src/+/afce9d93e76f2ff81baaa088a4ea25f67d1a76b3%5E%21/): previously it happened before `textInput`, now after `textInput`. As a consequence of this, `change` won't be fired if it is aborted while in composition 😕.

**Comment by suhaotian** [comment]
https://github.com/suhaotian/react-input maybe help someone

**Comment by eyesofkids** [comment]
There is a tricky solution. To call the handlechange after `compositionend` is fired.

```js
handleComposition  = (event) => {

    if(event.type === 'compositionend'){
      onComposition = false
      
      //fire change method to update for Chrome v53
      this.handleChange(event)
    
    } else{
      onComposition = true
    }
  }
```

check the demo here: [https://jsfiddle.net/eyesofkids/dcxvas28/11/](https://jsfiddle.net/eyesofkids/dcxvas28/11/)

**Comment by vincent714** [comment]
@chenxsan did you find out the solution?
you can detect the compositionStart and let a variable equal to true.
Then to use the variable, which you set, at onChange to see if it should fire the query

**Comment by eyesofkids** [comment]
I have submit a new issue for controlled components in [#8683 ](#8683)

The temporary solution for uncontrolled and controlled components(input, textarea) is upload to [react-compositionevent](https://github.com/eyesofkids/react-compositionevent).

**Comment by yesmeck** [comment]
Let's make https://github.com/facebook/react/pull/8438 happen.

**Comment by eyesofkids** [comment]
@yesmeck I saw the test only focus on the Webkit, it should be separate into Chrome and Safari because Chrome change its `compositionend` event triggered order after 53+.

**Comment by yesmeck** [comment]
@eyesofkids  Added a new  test case for Chrome under 53.

**Comment by SimeonC** [comment]
Just to add fuel to the fire, I've been trying to workaround this issue and discovered that the current version of iOS safari doesn't trigger the `compositionend` event when using Japanese Hiragana IME, I think this is intentional as the composition menu never seems to be closed.
On @eyesofkids example workaround the inputValue is never updated, though for me https://github.com/zhaoyao91/react-optimistic-input fixes the issue with Japanese IME.

**Comment by aprilandjan** [comment]
For anyone looking for solution for this, here's a ready-to-use component. <https://github.com/aprilandjan/react-starter/blob/test/search-input/src/components/SearchInput.js> Just use it instead of normal text input element and everything is ok.

**Comment by PokimLee** [comment]
@zhaoyao91  your [workaround](https://github.com/zhaoyao91/react-optimistic-input) just works!  thx a lot.

**Comment by cristianroche** [comment]
hey guys some news in this issue?

**Comment by craigkovatch** [comment]
@sophiebits help me/us understand how "make international input work properly in React" hasn't been a high priority for close to three years? This is pretty fundamental.

**Comment by sophiebits** [comment]
It hasn't been a high priority because onChange firing too frequently rarely causes problems. Where is it causing problems in your app?

**Comment by craigkovatch** [comment]
@sophiebits sorry accidentally clicked the 'X'. This can degrade performance if there are filtering operations or server callbacks involved. The approach shown in https://github.com/facebook/react/issues/3926#issuecomment-316049951 is a fine workaround but it seems like it would be simple (famous last words, I know) to just bake this into React. Seems like some in this thread have tried but found the internals a bit complex -- but maybe an engineer on your team could make quick work of it?

**Comment by gaearon** [comment]
Can somebody please help me understand: is the problem strictly in extra `onChange` calls *in the middle*? Or do you get incorrect value in the end?

**Comment by crochefluid** [comment]
@gaearon I'm gonna try it. Did you try that test on safari (mac/IOS)?

**Comment by gaearon** [comment]
It’s a Node test but it encodes sequences captured from different browsers and devices. Please see its source. You’d need to add sequences that are failing.

**Comment by yesmeck** [comment]
> So I suppose this issue is only about the extra onChange calls.

Exactly.

**Comment by robbyemmert** [comment]
I'm still getting this issue. It looks like this issue has been open for 3 years, does React support Chinese input in controlled components at the moment?

**Comment by cbengtson85** [comment]
Also seeing this in Japanese with certain characters...

**Comment by robbyemmert** [comment]
Here's a code sandbox that reproduces my issue. Looks like it's related to forms. Using inputs directly is fine.

https://codesandbox.io/s/0m1760xqnl

**Comment by robbyemmert** [comment]
I added some cases:
Using react state and plain inputs is fine
Using react state, plain forms, and plain inputs is fine
We're using a context-based form component which isn't working. It might be a context-related issue.

**Comment by robbyemmert** [comment]
Problem solved: I got it working in codepen. For some reason passing 'input' in as a component worked, when passing (props) => <input {...props]/> did not. 

Anyone have an idea what the difference is?

**Comment by otakustay** [comment]
It seems to cause incorrect result when IME is enabled

![e84721f3ec71a5ce043ef8290](https://user-images.githubusercontent.com/639549/53173667-16748100-3623-11e9-8f21-c6687d6d1f4b.gif)

**Comment by knubie** [comment]
I've experienced the same issue as @otakustay 
It seems impossible to support controlled input with IME input. I've traced the sequence of events to the following.

1. User types a letter, say `w`
1. onChange is triggered
1. State is updated with new value
1. New value is propagated down to the `input` by way of the `value` attribute.
1. The IME "composition" is interrupted at this point
   - There is a `w` string in the input element
   - There is also a separate `w` string stored in the IME buffer
1. The user types another letter, say `a`
1. The string in the input `a` combines with the string the IME buffer to produce `wwa`.
1. Repeat steps 1-7 to get a bunch of duplicate characters.

I've noticed that the bug only occurs if the input re-renders >15ms after the compositionUpdate event.

Right now my only solution is to switch away from controlled inputs.

**Comment by blessonbabu** [comment]
Any updates on this?

**Comment by huawin03** [comment]
update ??

**Comment by turutosiya** [comment]
Any update on this?

**Comment by eugle** [comment]
Stunned, I was confronted with this question

**Comment by cpdyj** [comment]
Interesting, looks the problem is not only about the multi-times onChange. If we not setState between the `onCompositionStart` and `onCompositionEnd`, the react will "control" the value as is. This action will interrupt the composition. That means we will not get the `onCompositionEnd` event......(If I'm wrong mention me plz.) But we can only change the state immediately(Otherwise we'll have to face the problem @knubie mentions). A reproduction at here(Looks like a "half-controlled" component): https://gist.github.com/cpdyj/6567437d96c315e9162778c8efdfb6e8

**Comment by craigkovatch** [comment]
> @hellendag I think we should not fire onChange until the IME string is committed.
I don't think this is a valid solution as a component might want to know the "uncommitted" IME string for e.g. filtering options in a list as the user types.

I am not sure if the approach I use in this other thread might help those hitting this issue, but here's a link just in case: https://github.com/facebook/react/issues/13104#issuecomment-691393940

**Comment by kazukinagata** [comment]
any updates?

**Comment by RebaekSoparkKitchen** [comment]
> any updates?

Hi, I face the same issue when I want to do some operation in handleChange function, and my solution is just use composition event to handle it.

If it works for you, extract it to hooks and we use reuse it easily.
```javascript
import React, { useState } from 'react';

export default function Solution() {
  const [value, setValue] = useState('');
  const [onComposition, setOnComposition] = useState(false);

  function handleComposition(event) {
    if (event.type === 'compositionstart') {
      setOnComposition(true);
    }
    if (event.type === 'compositionend') {
      setOnComposition(false);
    }
  }
  function handleChange(e) {
    let temp = e.target.value;
    if (!onComposition) {
      // any operation you want to do...
      temp = temp.replace(/z/g, 'Z');
    }
    setValue(temp);
  }

  return (
    <input
      onChange={handleChange}
      value={value}
      onCompositionStart={handleComposition}
      onCompositionEnd={handleComposition}
    />
  );
}
```

**Comment by krmao** [comment]
also this code [https://zhuanlan.zhihu.com/p/415365117](https://zhuanlan.zhihu.com/p/415365117)

**Comment by eyesofkids** [comment]
> also this code https://zhuanlan.zhihu.com/p/415365117

```jsx
import { useRef, forwardRef } from 'react'

function InputIME(props, ref) {
  const { onChange, ...otherProps } = props
  // log if on composition
  const onComposition = useRef(false)

  const _onChange = (event) => {
    // IME method start
    if (event.type === 'compositionstart') {
      onComposition.current = true
      return
    }

    // IME method end
    if (event.type === 'compositionend') {
      onComposition.current = false
    }

    // handle onChange
    if (!onComposition.current) {
      onChange(event)
    }
  }

  return (
    <input
      {...otherProps}
      ref={ref}
      onChange={_onChange}
      onCompositionEnd={_onChange}
      onCompositionStart={_onChange}
    />
  )
}

export default forwardRef(InputIME)
```

**Comment by kamikat** [comment]
Another hook-ish workaround:

```ts
import { useEffect, useState, useRef } from "react";

export function useAsyncInputRef<T extends HTMLElement & { value: string }>(
  value: string,
  forwardRef?: React.RefObject<T>
) {
  const [compositionFlag, setComposition] = useState(false);
  const newRef = useRef<T>(null);
  const ref = forwardRef || newRef;

  useEffect(() => {
    const el = ref.current;
    if (el) {
      const compositionStart = () => setComposition(true);
      const compositionEnd = () => setComposition(false);
      el.addEventListener("compositionstart", compositionStart);
      el.addEventListener("compositionend", compositionEnd);
      return () => {
        el.removeEventListener("compositionstart", compositionStart);
        el.removeEventListener("compositionend", compositionEnd);
        setComposition(false);
      };
    }
  }, [ref]);

  useEffect(() => {
    if (ref.current && compositionFlag === false) {
      ref.current.value = value;
    }
  }, [compositionFlag, value, ref]);

  return ref;
}
```

```tsx
render(
  <input
    type="text"
    ref={useAsyncInputRef(value)}
    onChange={(e) => setValue(e.target.value)}
  />
);
```

Live demo: https://codepen.io/kamikat/pen/OJwMJLr

**Comment by zheeeng** [comment]
Is there an official improvement plan for this? I'm suffering from using workarounds anywhere.

**Comment by weihong1028** [comment]
Nice,thank you！                               ——weihong

**Comment by lethi2k** [comment]
import React, { useState } from 'react';

export default function Solution() {
  const [onComposition, setOnComposition] = useState(false);

  function handleComposition(event) {
    if (event.type === 'compositionstart') {
      setOnComposition(true);
    }
    if (event.type === 'compositionend') {
      setOnComposition(false);
    }
  }
  function onPressEnter(e) {
    if (!onComposition) {
       // code here
    }
  }

  return (
    <input
      onPressEnter={onPressEnter}
      value={value}
      onCompositionStart={handleComposition}
      onCompositionEnd={handleComposition}
    />
  );
}

**Comment by weihong1028** [comment]
Nice,thank you！                               ——weihong
